### PR TITLE
Add support for remote operation cache requests

### PR DESCRIPTION
**Pull request by beweedon**
Cache requests in remote operations behave slightly differently than in
classic UIA. Rather than being tied to a specific operation (e.g.
GetParentElementBuildCache), a remote element's cache can be populated
at any time. This is because we are already in the remote process, so we
don't need to attach a cache request to an outgoing operation to reduce
overall latency.

The instructions IsCacheRequest, NewCacheRequest, PopulateCache,
CacheRequestAddProperty, and CacheRequestAddPattern are added. The WinRT
API provides a fairly thin wrapper on top of them.

The UiaOperationAbstraction library is a bit more complex. Since it
should work with both remote and classic UIA, we need to follow a
similar interface. This is done by introducing the UiaCacheRequest type.
This type cannot be imported or exported from a remote operation, but it
can be constructed from either an IUIAutomationCacheRequest or an
AutomationRemoteCacheRequest. In methods that accept cache requests, we
then check if we're in a remote operation. If we are, we call
PopulateCacheRequest using the cache request's remote form. If we're not
remote, we retrieve the underlying COM pointer and pass it on to UIA as
usual.

### CODE REVIEW

**Comment by beweedon** [comment]
I haven't finished running tests yet. I'll resolve this comment once that's done.

File: src/UIAutomation/FunctionalTests/FunctionalTests.vcxproj [hunk 1]
```diff
@@ -93,12 +93,12 @@
...
       <WarningLevel>Level4</WarningLevel>
       <SDLCheck>true</SDLCheck>
-      <AdditionalIncludeDirectories>$(VCInstallDir)UnitTest\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(VCInstallDir)UnitTest\include;$(SolutionDir)UiaOperationAbstraction;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <PreprocessorDefinitions>WIN32;_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <UseFullPaths>true</UseFullPaths>
       <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
       <LanguageStandard>stdcpp17</LanguageStandard>
-      <ConformanceMode>true</ConformanceMode>
+      <ConformanceMode>false</ConformanceMode>
```
Comment (mlalic) [inline src/UIAutomation/FunctionalTests/FunctionalTests.vcxproj:101]: >false [](start = 23, length = 5)

It's unfortunate this change is needed, since the WinRT API should be `permissive-` conformant.

Regardless, it might make sense to have the abstraction and WinRT APIs separate anyway? They kind of are separate projects too...

Anyway, not a big deal or a blocker in any way for this PR, just thinking outloud.

File: src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp [hunk 2]
```diff
@@ -0,0 +1,471 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CppUnitTest.h"
+
+#include "ModernApp.h"
+#include "TestUtils.h"
+
+namespace WinRTWrapperTests
```
Comment (mlalic) [inline src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp:10]: >WinRTWrapperTests [](start = 10, length = 17)

I'm trying to call this the "Builder" library. I think that's a bit better than wrapper, but I'm also open to ideas. :)

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    UiaOperationAbstraction::UiaCacheRequest CreateCacheRequest(const bool useRemoteOperations)
+    {
+        if (useRemoteOperations)
+        {
+            UiaOperationScope::GetCurrentDelegator()->NewRemoteCacheRequest();
+        }
```
Comment (mlalic) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:48]: Hmm, maybe I'm missing something here, but this feels like it should be part of the abstraction library itself.

The same way how we don't need users to manually figure out how to initialize a `UiaArray`, we should have `UiaCacheRequest` initialize itself correctly too. Then we wouldn't need this `NewRemoteCacheRequest` method--again, same as for `UiaArray`.

Could this work?

File: src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl [hunk 4]
```diff
@@ -560,6 +574,52 @@ namespace Microsoft.UI.UIAutomation
...
+        Toggle = 10015,
+        Transform = 10016,
+        ScrollItem = 10017,
+        LegacyIAccessible = 10018,
+        ItemContainer = 10019,
+        VirtualizedItem = 10020,
+        SynchronizedInput = 10021,
+        ObjectModel = 10022,
+        Annotation = 10023,
+        UIA_TextPattern2Id = 10024,
```
Comment (mlalic) [inline src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl:603]: >UIA_TextPattern2Id [](start = 8, length = 18)

There are a few enum variants that different than the others :)

(Seems like maybe the ones that have a number in the name?)

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            auto name = element.GetName(false /*useCachedApi*/);
+            scope.BindResult(name);
+
+            scope.Resolve();
+
+            Assert::AreEqual(std::wstring(static_cast<wil::shared_bstr>(name).get()), std::wstring(L"Display is 0"));
+        }
+        TEST_METHOD(ElementGetNameLocalTest)
+        {
+            GetParentElementTest(false);
```
Comment (csauls) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:90]: ElementGetNameTest

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        TEST_METHOD(ElementGetNameRemoteTest)
+        {
+            GetParentElementTest(true);
```
Comment (csauls) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:95]: ElementGetNameTest

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    template<class T>
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    UiaOperationAbstraction::UiaCacheRequest CreateCacheRequest(const bool useRemoteOperations)
+    {
+        if (useRemoteOperations)
+        {
+            UiaOperationScope::GetCurrentDelegator()->NewRemoteCacheRequest();
```
Comment (csauls) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:47]: Need to return the object, right?

File: src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp (hunk 2 above)
Comment (mwcampbell) [reply src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp:10]: I also prefer "builder".

File: src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl (hunk 4 above)
Comment (mwcampbell) [reply src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl:603]: It must be the fact that the name doesn't end with `PatternId`. I'm pretty sure I did a special case for this somewhere else in the code generator.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            });
+
+            auto scope = UiaOperationScope::StartNew();
+
+            UiaElement element = calc;
+            UiaTextPattern textPattern = element.GetTextPattern(false /*useCached*/);
+            UiaTextRange textRange = textPattern.GetDocumentRange();
+
+            auto cacheRequest = CreateCacheRequest(useRemoteOperations);
+            cacheRequest.AddProperty(static_cast<winrt::AutomationPropertyId>(UIA_NamePropertyId));
```
Comment (LukaszMakar) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:120]: If possible, I would add a mirroring test for retrieving and using cached patterns.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h [hunk 9]
```diff
@@ -1907,4 +1913,16 @@ namespace UiaOperationAbstraction
     };
 
     bool ShouldUseRemoteApi();
+
+    // Each of these helpers is an overload which populates the cache(s) for
```
Comment (LukaszMakar) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:1917]: **Suggestion:** You could make it explicit and differentiate by adding appropriate suffixes:
1. `PopulateCacheForElement`,
2. `PopulateCacheForElementArray` or `PopulateCacheForArrayElements`

File: src/UIAutomation/FunctionalTests/FunctionalTests.vcxproj (hunk 1 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/FunctionalTests.vcxproj:101]: Yeah, the issue I ran into that required this was that UiaBool was being used in the abstraction headers before it was defined. I'm not sure if there was something else that also required conformance mode false, but I didn't dig too deep.

File: src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp (hunk 2 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/WinRTWrapperTests.cpp:10]: Sounds good to me.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 8 above)
Comment (LukaszMakar) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:121]: Never mind. It's in the other test suite.

---
In reply to: [459171231](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459171231) [](ancestors = 459171231)

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 3 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:48]: Yeah, I think you're right. Now that I'm thinking about it though it also raises the question of what are "supported" ways to create a cache request. In theory, we shouldn't allow it to be constructed from an IUIAutomationCacheRequest, since that cache request may contain fields (i.e. tree scope, etc.) which we don't support. It allows a weird backdoor to get around the restrictions on the UiaCacheRequest type.

I'll think about that some more.

File: src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl (hunk 4 above)
Comment (beweedon) [reply src/UIAutomation/Microsoft.UI.UIAutomation/Microsoft.UI.UIAutomation.idl:603]: Lol so there are. Thanks for checking this. I'll fix it up.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 8 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:121]: Fair. It wouldn't hurt to have extra coverage on that.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h (hunk 9 above)
Comment (beweedon) [reply src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:1917]: So, the reason why I have the overload is that the functions that call the helpers are generated by the code gen. Using overloads seemed cleaner than switching over possible types in the code generator.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 8 above)
Comment (LukaszMakar) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:121]: That covers the other of the 2 functionalities behind this new feature.

And you have most of the pieces here already (`Text` pattern + getting the document range + getting text + comparing the text would do just fine for the test imho).

---
In reply to: [459176822](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459176822) [](ancestors = 459176822)

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h (hunk 9 above)
Comment (LukaszMakar) [reply src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:1917]: Got'cha. Fine by me. :)

---
In reply to: [459177495](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459177495) [](ancestors = 459177495)

**Comment by beweedon** [comment]
Marko and I both ran into an issue where it seems UiaOperationAbstraction is being built before the WinRT library, which is causing a build failure if you manually clear all build artifacts. Looking into this now.

File: src/UIAutomation/FunctionalTests/TestUtils.h [hunk 10]
```diff
@@ -18,3 +40,49 @@ void LogOutput(Args&&... args)
...
+    THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+
+    for (int i = 0; i < retries; ++i)
+    {
+        winrt::com_ptr<IUIAutomationElement> element;
+        const auto getFocusedElementHr = automation->GetFocusedElement(element.put());
+        if (getFocusedElementHr == E_INVALIDARG)
+        {
+            LogOutput(L"GetFocusedElement returned E_INVALIDARG; retrying");
+            Sleep(50);
```
Comment (paul-olczak) [inline src/UIAutomation/FunctionalTests/TestUtils.h:58]: Consider making the 50 here and below a named constant

**Comment by mwcampbell** [comment]
In the UIA operation abstraction tests, when you add a property or pattern to the cache request, you shouldn't need to cast it to the WinR version of the enum. `UiaEnum` takes care of that.

**Comment by mwcampbell** [comment]
Unless I'm misunderstanding `CacheRequestNoReturnTest`, I don't think its name and description are accurate. After all, the window element, where the cache is populated, _is_ returned as a result.

**Comment by mwcampbell** [comment]
In the abstraction, I don't think the definition of `UiaCacheRequest` belongs in the generated code, because it can be completely hand-written. I think I know why it's defined in the generated header file; it depends on the enum types that are defined in that header file. But I think this just means we need a separate generated fragment for the enum definitions, which can be included right after the definition of `UiaEnum` itself. Then, if you define `UiaCacheRequest` in UiaOperationAbstraction.h, you don't have to make the code generator write that definition, and you can add usability improvements like a default constructor.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp [hunk 11]
```diff
@@ -654,6 +654,23 @@ namespace UiaOperationAbstraction
...
+        {
+            return m_remoteOperation.NewCacheRequest();
+        }
+        else
+        {
+            winrt::com_ptr<IUIAutomation> automation;
+            THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+
+            winrt::com_ptr<IUIAutomationCacheRequest> cacheRequest;
+            automation->CreateCacheRequest(cacheRequest.put());
```
Comment (LukaszMakar) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:669]: >CreateCacheRequest [](start = 24, length = 18)

`THROW_IF_FAILED`

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp [hunk 12]
```diff
@@ -654,6 +654,23 @@ namespace UiaOperationAbstraction
...
+    {
+        if (m_useRemoteApi)
+        {
+            return m_remoteOperation.NewCacheRequest();
+        }
+        else
+        {
+            winrt::com_ptr<IUIAutomation> automation;
+            THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+
```
Comment (LukaszMakar) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:667]: You don't need it.

The abstraction relies on having one, pre-set instance of `IUIAutomation`:
```
wil::object_without_destructor_on_shutdown<wil::com_ptr<IUIAutomation>> g_automation;
```

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            ElementGetNameTest(false);
+        }
+
+        TEST_METHOD(ElementGetNameRemoteTest)
+        {
+            ElementGetNameTest(true);
+        }
+
+        // Asserts that GetEnclosingElement can be called with a cache request.
+        void GetEnclosingElementTest(const bool useRemoteOperations)
```
Comment (LukaszMakar) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:82]: >GetEnclosingElementTest [](start = 13, length = 23)

I would rename the test and test helper functions to reflect what they really test.

(I was about to write a comment in the last test in the file that it is important to test the case where `Text` pattern is available only to realize that this test does it :))

File: src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstractionImpl.g.cpp [hunk 14]
```diff
@@ -5020,6 +5030,93 @@
...
+    UiaCacheRequest::UiaCacheRequest(_In_ IUIAutomationCacheRequest* cacheRequest):
+        UiaTypeBase(MakeWinrtComPtr(cacheRequest))
+    {
+    }
+
+    UiaCacheRequest::UiaCacheRequest(winrt::com_ptr<IUIAutomationCacheRequest> const& cacheRequest):
+        UiaTypeBase(cacheRequest)
+    {
+    }
+
```
Comment (LukaszMakar) [inline src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstractionImpl.g.cpp:5042]: For my understanding, I think the discussion around importing cache requests ended up on not needing that operation. And due to that, there is no `ToRemove()` call (and no supporting infrastructure to convert to remote).

From the practical point of view -- unless I am missing something -- this means that we should switch to using this (abstraction) type instead of `IUIAutomationCacheRequest` whenever we work with `UiaOperationAbstraction` (regardless of whether operations are executed locally or remotely) as otherwise:
1. The cache request will always be seen as empty/null.
2. Adding anything to the underlying cache request (when in the "remote" mode) will result in exceptions due to `std::get` finding out that the type of the cache request is not remote.

Is that correct to say? Or is there something I am missing? I am trying to understand the shift in paradigm of usability.

Previously, whomever used the abstraction did not really have to think twice about how abstraction types have been created. This sounds different, so I thought I would ask to understand the piece better. :)

**Comment by beweedon** [comment]
> Unless I'm misunderstanding `CacheRequestNoReturnTest`, I don't think its name and description are accurate. After all, the window element, where the cache is populated, _is_ returned as a result.

Huh, thanks! I had it returning the value for easy copy-pasting while writing the tests, but I guess I forgot to delete it.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+            });
+
+            auto scope = UiaOperationScope::StartNew();
+
+            UiaElement element = calc;
+            UiaElement childText = element.GetFirstChildElement().GetFirstChildElement();
+            UiaTextPattern textPattern = childText.GetTextPattern(false /*useCached*/);
+            UiaTextRange textRange = textPattern.GetDocumentRange();
+
+            auto cacheRequest = UiaOperationScope::GetCurrentDelegator()->CreateCacheRequest();
```
Comment (mlalic) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:103]: >CreateCacheRequest [](start = 74, length = 18)

I still don't understand why we need a separate helper function for creating a `UiaCacheRequest` but we don't need one for creating a `UiaArray`.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 13 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:82]: True, they should probably mention cache requests in the name :). The reason why I have this test and the GetParentElement test is that these two functions handle cache requests slightly differently, and I wanted coverage across them. I could maybe also rename these tests to CacheRequestPatternMethodTest and CacheRequestNavigationMethodTest.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 3 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp]: I'm using this thread to reply generally to the multiple comments by @mlalic, @mwcampbell, and @LukaszMakar. I was thinking about how to create a UiaCacheRequest overnight, and I believe the best approach is probably a default constructor.

Allowing a constructor that accepts IUIAutomationCacheRequest doesn't make sense, since that local type may have extra state (such as tree scope) which we don't support. It also just isn't really necessary to have an AutomationRemoteCacheRequest constructor since it doesn't need to be used.

We basically always need to create an empty cache request which determines its remoteness based on the current delegator. It needs to be initialized local or remote and never change. This is slightly different from UiaArray because UiaArray can be instantiated without regard for the remoteness of the current operation. UiaCacheRequest can't change remoteness, so it needs to be initialized correctly.

tl;dr I think the solution is to make the only way to construct a UiaCacheRequest a default initializer that reads the current delegator's remoteness value.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp (hunk 11 above)
Comment (beweedon) [reply src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:669]: @mwcampbell 
> In the abstraction, I don't think the definition of `UiaCacheRequest` belongs in the generated code, because it can be completely hand-written. I think I know why it's defined in the generated header file; it depends on the enum types that are defined in that header file. But I think this just means we need a separate generated fragment for the enum definitions, which can be included right after the definition of `UiaEnum` itself. Then, if you define `UiaCacheRequest` in UiaOperationAbstraction.h, you don't have to make the code generator write that definition, and you can add usability improvements like a default constructor.

I created it with the code generator because it provides boilerplate like IsNull and operators. Looking more closely at other types, though, you're probably right that it differs enough it's easier to hand-write it.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 15 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:103]: See above thread.

File: src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstractionImpl.g.cpp (hunk 14 above)
Comment (beweedon) [reply src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstractionImpl.g.cpp:5042]: See above thread for more details, but yes, you're correct :). We shouldn't allow constructing using IUIAutomationCacheRequest.

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp (hunk 11 above)
Comment (beweedon) [reply src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.cpp:669]: Huh, how did this get here? Weird.

Anyway, @mwcampbell, this is meant to reply to the quoted comment lol.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 3 above)
Comment (mlalic) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp]: This reasoning sounds sensible to me. I also like doing the simpler thing (just the one ctor) until there's a clear need for more, especially when from everything that we see right now, we can't find a good reason anyway.

---
In reply to: [459790094](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459790094) [](ancestors = 459790094)

Comment (LukaszMakar) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp]: Agreed. I expressed my concerns in another comment. This feels fine to me. :)

---
In reply to: [459798282](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459798282) [](ancestors = 459798282,459790094)

File: src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h [hunk 16]
```diff
@@ -964,6 +973,8 @@ namespace UiaOperationAbstraction
         }
     };
 
+#include "UiaTypeAbstractionEnums.g.h"
```
Comment (beweedon) [inline src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:976]: This is probably fine for this PR, but I think we need to restructure UiaOperationAbstraction in the future. I briefly experimented with having two sets of files: one with only basic types (i.e. types which don't depend on other non-basic abstraction types) and one of non-basic types. Each of these sets would have generated files and un-generated files. I didn't fully test this so I'm not sure if there would be other compilation errors related to needing to keep generated and un-generated code separate. But the current structure makes it difficult to make changes without interleaving header/cpp files, so I feel like a refactoring would have a lot of benefits.

I suppose this also ties in with the current discussion about the code generator.

File: src/UIAutomation/FunctionalTests/TestUtils.h [hunk 17]
```diff
@@ -18,3 +48,50 @@ void LogOutput(Args&&... args)
...
+{
+    static constexpr DWORD c_sleepMilliseconds = 50;
+    winrt::com_ptr<IUIAutomation> automation;
+    THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+
+    for (int i = 0; i < retries; ++i)
+    {
+        winrt::com_ptr<IUIAutomationElement> element;
+        const auto getFocusedElementHr = automation->GetFocusedElement(element.put());
+        if (getFocusedElementHr == E_INVALIDARG)
```
Comment (LukaszMakar) [inline src/UIAutomation/FunctionalTests/TestUtils.h:64]: When is this true?

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+    {
+    public:
+        // Asserts that you can get the name of a UiaElement.
+        void ElementGetNameTest(const bool useRemoteOperations)
+        {
+            ModernApp app(L"Microsoft.WindowsCalculator_8wekyb3d8bbwe!App");
+            app.Activate();
+            auto calc = WaitForElementFocus(L"Display is 0");
+
+            winrt::com_ptr<IUIAutomation> automation;
```
Comment (LukaszMakar) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:53]: Would you mind adding a helper for it that returns the cleanup guard? The entire initialization-uninitialization block is common for all the tests here.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+            THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+            UiaOperationAbstraction::Initialize(useRemoteOperations, automation.get());
+            auto cleanup = wil::scope_exit([&]()
```
Comment (LukaszMakar) [inline src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:56]: Doesn't need any concrete capture since it just calls the global cleanup function.

File: src/UIAutomation/FunctionalTests/TestUtils.h (hunk 17 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/TestUtils.h:64]: This is copied from the deleted FunctionalTests.cpp, which I assume is copied from CommonUtil.cpp, which has a comment explaining this. @mlalic, is there a reason you removed the comment?

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 18 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:53]: Absolutely. I was considering it before and I'm glad you're giving me an excuse to do it now :)

File: src/UIAutomation/FunctionalTests/TestUtils.h (hunk 17 above)
Comment (beweedon) [reply src/UIAutomation/FunctionalTests/TestUtils.h:64]: Resolved offline.

File: src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp (hunk 19 above)
Comment (mlalic) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:56]: This is fine as it is. It's always safe to use a blanket by-ref capture with `scope_exit` and there's no need to think about what to capture all the time; just consistently use by ref and it's always correct.

---
In reply to: [460283878](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r460283878) [](ancestors = 460283878)

Comment (LukaszMakar) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:56]: Discussed offline: Let's not do it. :)

Comment (beweedon) [reply src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp]: When moving this to a function I got rid of the capture, since when returning a `wil::scope_exit` from a function it definitely shouldn't reference local state. Thanks for the feedback :)
